src/Xamarin.Android.Build.Tasks/Tasks/JavaSourceUtils.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (C) 2012 Xamarin, Inc. All rights reserved.
+
+using System;
+using System.Linq;
+using System.IO;
+using System.Reflection;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Xamarin.Android.Tools;
+
+namespace Xamarin.Android.Tasks
+{
+	public class JavaSourceUtils : AndroidToolTask
+	{
+		public  override    string  TaskPrefix => "JSU";
+
+		[Required]
+		public  string      MonoAndroidToolsDirectory   { get; set; }
+
+		[Required]
+		public  string      JavaSdkDirectory            { get; set; }
+
+		[Required]
+		public  ITaskItem   OutputDirectory             { get; set; }
+
+		[Required]
+		public  ITaskItem[] InputFiles                  { get; set; }
+
+		public  ITaskItem[] References                  { get; set; }
+
+		public  ITaskItem[] BootClassPath               { get; set; }
+
+		public  ITaskItem   JavadocCopyrightFile        { get; set; }
+		public  string      JavadocUrlPrefix            { get; set; }
+		public  string      JavadocUrlStyle             { get; set; }
+
+		public  string      JavaOptions                 { get; set; }
+
+		public  string      JavaMaximumHeapSize         { get; set; }
+
+		[Output]
+		public  ITaskItem   OutputJavadocXml            { get; set; }
+
+		string  responseFilePath;
+
+		public override bool RunTask ()
+		{
+			if (InputFiles == null || InputFiles.Count () == 0) {
+				Log.LogCodedError ("XA1020", Properties.Resources.XA1020);
+				return false;
+			}
+
+			if (References != null)
+				foreach (var path in References)
+					if (!Directory.Exists (path.ItemSpec) && !File.Exists (path.ItemSpec))
+						Log.LogCodedError ("XA1022", Properties.Resources.XA1022, path.ItemSpec);
+
+			if (Log.HasLoggedErrors)
+				return false;
+
+			OutputJavadocXml	= new TaskItem (Path.Combine (OutputDirectory.ItemSpec, GetOutputFileName (InputFiles)) + ".xml");
+
+			try {
+				return base.RunTask ();
+			}
+			finally {
+				File.Delete (responseFilePath);
+			}
+		}
+
+		static string GetOutputFileName (ITaskItem[] items) =>
+			Files.HashString (
+				string.Join ("\n", items.Select (item => Path.GetFullPath (item.ItemSpec).Replace ('\\', '/'))));
+
+
+		protected override string GenerateCommandLineCommands ()
+		{
+			responseFilePath    = CreateResponseFile ();
+
+			var cmd = new CommandLineBuilder ();
+
+			// Add the JavaOptions if they are not null
+			// These could be any of the additional options
+			if (!string.IsNullOrEmpty (JavaOptions)) {
+				cmd.AppendSwitch (JavaOptions);
+			}
+
+			if (!string.IsNullOrEmpty (JavaMaximumHeapSize)) {
+				cmd.AppendSwitchIfNotNull("-Xmx", JavaMaximumHeapSize);
+			}
+
+			// Arguments sent to java.exe
+			cmd.AppendSwitchIfNotNull ("-jar ", Path.Combine (MonoAndroidToolsDirectory, "java-source-utils.jar"));
```

Comment:
Do we want to hard code this `java-source-utils.jar` ? Or have it as a property like we do for `ToolName` in other tasks?

Reply:
We should follow the other tasks.